support/ebpf/beam_tracer.ebpf.c

```diff
@@ -0,0 +1,103 @@
+#include "bpfdefs.h"
+#include "tracemgmt.h"
+#include "types.h"
+
+// The number of frames to unwind per frame-unwinding eBPF program.
+#define FRAMES_PER_PROGRAM   16
+
+bpf_map_def SEC("maps") beam_procs = {
+  .type = BPF_MAP_TYPE_HASH,
+  .key_size = sizeof(pid_t),
+  .value_size = sizeof(BEAMProcInfo),
+  .max_entries = 256,
+};
+
+static EBPF_INLINE ErrorCode unwind_one_beam_frame(PerCPURecord *record, BEAMProcInfo *info, bool top) {
+  UnwindState *state = &record->state;
+  Trace *trace = &record->trace;
+  u64 sp = state->sp, fp = state->fp, pc = state->pc;
+
+  DEBUG_PRINT("beam: pc: %llx, sp: %llx, fp: %llx", pc, sp, fp);
+
+  bpf_probe_read_user(&state->fp, sizeof(u64), (void*)fp);
+  bpf_probe_read_user(&state->pc, sizeof(u64), (void*)(fp+8));
+  bpf_probe_read_user(&state->sp, sizeof(u64), (void*)(fp+16));
+
+  if (fp && sp) {
+    unwinder_mark_nonleaf_frame(state);
+  }
+
+  _push_with_return_address(trace, info->active_ranges, pc, FRAME_MARKER_BEAM, state->return_address);
+
+  if (state->fp) {
+    unwinder_mark_nonleaf_frame(state);
+  }
+
+  return ERR_OK;
+}
+
+// unwind_beam is the entry point for tracing when invoked from the native tracer
+// or interpreter dispatcher. It does not reset the trace object and will append the
+// BEAM stack frames to the trace object for the current CPU.
+static EBPF_INLINE int unwind_beam(struct pt_regs *ctx) {
+  PerCPURecord *record = get_per_cpu_record();
+  if (!record) {
+    return -1;
+  }
+
+  Trace *trace = &record->trace;
+  u32 pid = trace->pid;
+
+  DEBUG_PRINT("==== unwind_beam %d ====", trace->stack_len);
+
+  int unwinder = PROG_UNWIND_STOP;
+  ErrorCode error = ERR_OK;
+  BEAMProcInfo *info = bpf_map_lookup_elem(&beam_procs, &pid);
+
+  if (!info || !info->the_active_code_index || !info->r) {
```

Comment:
If `the_active_code_index` or `r` cannot be determined, `info` should not exist. The error checking should be on the go plugin side. This allows removing the checks here.